src/Platform/Vulkan/VulkanPipelineBuilder.cc

```diff
@@ -0,0 +1,372 @@
+                                  VK_NULL_HANDLE,
+                                  1,
+                                  &pipelineInfo,
+                                  nullptr,
+                                  &graphicsPipeline) != VK_SUCCESS)
+    {
+      throw std::runtime_error("failed to create graphics pipeline!\n");
+    }
+    else { ESP_CORE_INFO("Graphic pipeline created correctly"); }
+
+    vkDestroyShaderModule(VulkanDevice::get_logical_device(), m_fragment_shader_module, nullptr);
+    vkDestroyShaderModule(VulkanDevice::get_logical_device(), m_vertex_shader_module, nullptr);
+
+    m_is_fragment_shader_module = false;
+    m_is_vertex_shader_module   = false;
+    m_is_pipeline_layout        = false;
+
+    return std::unique_ptr<EspPipeline>{
+      new VulkanPipeline(m_pipeline_layout, graphicsPipeline, std::move(m_uniform_manager))
+    };
+  }
+} // namespace esp
+
+/* --------------------------------------------------------- */
+/* ------------------ HELPFUL FUNCTIONS -------------------- */
+/* --------------------------------------------------------- */
+
+static std::vector<char> read_file(const std::string& filename)
+{
+  std::ifstream file(filename, std::ios::ate | std::ios::binary);
+
+  if (!file.is_open())
+  {
+    ESP_CORE_ERROR("The file doesn't exist: {0}", filename);
+    throw std::runtime_error("The file doesn't exist.");
+  }
+
+  size_t fileSize = (size_t)file.tellg();
+  std::vector<char> buffer(fileSize);
+
+  file.seekg(0);
+  file.read(buffer.data(), fileSize);
+
+  file.close();
+
+  return buffer;
+}
+
+static VkShaderModule create_shader_module(const std::vector<char>& code, VkDevice device)
+{
+  VkShaderModuleCreateInfo createInfo{};
+  createInfo.sType    = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO;
+  createInfo.codeSize = code.size();
+  createInfo.pCode    = reinterpret_cast<const uint32_t*>(code.data());
+
+  VkShaderModule shaderModule;
+  if (vkCreateShaderModule(device, &createInfo, nullptr, &shaderModule) != VK_SUCCESS)
+  {
+    ESP_CORE_ERROR("Failed to create shader module");
+    throw std::runtime_error("Failed to create shader module");
+  }
+
+  return shaderModule;
+}
+
+static VkDescriptorSetLayoutBinding create_dslb(esp::VulkanMetaUniform& data)
+{
+  auto stage = VK_SHADER_STAGE_VERTEX_BIT;
+  if (data.m_stage == esp::EspUniformShaderStage::ESP_FRAG_STAGE) { stage = VK_SHADER_STAGE_FRAGMENT_BIT; }
+
+  switch (data.m_uniform_type)
+  {
+  case esp::EspUniformType::ESP_BUFFER_UNIFORM:
+  {
+    VkDescriptorSetLayoutBinding uboLayoutBinding{};
+    uboLayoutBinding.binding            = data.m_binding;
+    uboLayoutBinding.descriptorCount    = data.m_count_of_elements;
+    uboLayoutBinding.descriptorType     = VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER;
+    uboLayoutBinding.pImmutableSamplers = nullptr;
+    uboLayoutBinding.stageFlags         = stage;
+    return uboLayoutBinding;
+  }
+
+  case esp::EspUniformType::ESP_SMALL_FAST_UNIFORM:
```

Comment:
@DawidSzpejna I think "PUSH_CONSTANT" is more representative name than "SMALL_FAST_UNIFORM"

Reply:
With great regret, I must unfortunately acknowledge that there is no common ground on which I could agree with you on this matter. Due to the fact that their existence manifests exclusively within the confines of the volcano. That's why it is not universal approach to label them as `PUSH_CONSTANT` @sebastianSzarafin